datalad/local/remove.py

```diff
@@ -0,0 +1,229 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""High-level interface for removing dataset content
+
+"""
+
+__docformat__ = 'restructuredtext'
+
+import logging
+from os.path import lexists
+
+from datalad.core.local.save import Save
+from datalad.core.local.status import get_paths_by_ds
+from datalad.distributed.drop import Drop
+from datalad.distribution.dataset import (
+    datasetmethod,
+    require_dataset,
+    EnsureDataset,
+)
+from datalad.interface.base import (
+    Interface,
+    build_doc,
+)
+from datalad.interface.common_opts import (
+    jobs_opt,
+    save_message_opt,
+)
+from datalad.interface.utils import eval_results
+from datalad.support.constraints import (
+    EnsureChoice,
+    EnsureNone,
+    EnsureStr,
+)
+from datalad.support.gitrepo import GitRepo
+from datalad.support.param import Parameter
+from datalad.utils import (
+    ensure_list,
+    rmtree,
+)
+
+lgr = logging.getLogger('datalad.local.remove')
+
+
+@build_doc
+class Remove(Interface):
+    """Remove components from datasets
+
+    Removing "unlinks" a dataset component, such as a file or subdataset, from
+    a dataset. Such a removal advances the state of a dataset, just like adding
+    new content. A remove operation can be undone, by restoring a previous
+    dataset state, but might require re-obtaining file content and subdatasets
+    from remote locations.
+
+    This command relies on the 'drop' command for safe operation. By default,
+    only file content from datasets which will be uninstalled as part of
+    a removal will be dropped. Otherwise file content is retained, such that
+    restoring a previous version also immediately restores file content access,
+    just as it is the case for files directly committed to Git. This default
+    behavior can be changed to always drop content prior removal, for cases
+    where a minimal storage footprint for local datasets installations is
+    desirable.
+
+    Removing a dataset component is always a recursive operation. Removing a
+    directory, removes all content underneath the directory too. If
+    subdatasets are located under a to-be-removed path, they will be
+    uninstalled entirely, and all their content dropped. If any subdataset
+    can not be uninstalled safely, the remove operation will fail and halt.
+    """
```

Comment:
Maybe this is the right place to mention that remove won't operate on ``.``?

Reply:
Not sure. My view was that removing CWD would be extremely unexpected behavior -- hence confirming that it won't happen is not needed. Same rational as for UNIX `rm`. It is also not in the docs that it won't remove CWD, but still wouldn't.

Reply:
ok